bpo-32904: Fix a potential crash in os.chdir() and os.getcwd() on Windows

izbyshev opened the pull request:
win32_wchdir() retries GetCurrentDirectory() with a larger buffer
if necessary, but doesn't check whether the new buffer is large enough.
Another thread could change the current directory in meanwhile,
so the buffer could turn out to be still not large enough, left in
an uninitialized state and passed to SetEnvironmentVariable() afterwards.

A similar issue occurs in posix_getcwd().

<!-- issue-number: bpo-32904 -->
https://bugs.python.org/issue32904
<!-- /issue-number -->

eryksun commented on Modules/posixmodule.c:
```diff
@@ -1523,11 +1523,12 @@ win32_wchdir(LPCWSTR path)
-        result = GetCurrentDirectoryW(result, new_path);
-        if (!result) {
+        int retry_result = GetCurrentDirectoryW(result, new_path);
+        if (!retry_result || retry_result > result) {
```
Changing the current directory shouldn't fail because of a race condition. I suggest using a while loop here (untested):

    static BOOL __stdcall
    win32_wchdir(LPCWSTR path)
    {
        wchar_t path_buf[MAXPATHLEN], *new_path = path_buf;
        int result, length = Py_ARRAY_LENGTH(path_buf);
        wchar_t env[4] = L"=x:";

        if(!SetCurrentDirectoryW(path))
            return FALSE;

        while (1) {
            wchar_t *temp;
            result = GetCurrentDirectoryW(length, new_path);
            if (!result)
                goto exit;
            if (result <= length)
                break;
            length = result;
            temp = PyMem_RawRealloc(new_path != path_buf ? new_path : NULL,
                length * sizeof(wchar_t));
            if (!temp)
                goto exit;
            new_path = temp;
        }
        
        if (wcsncmp(new_path, L"\\\\", 2) == 0 ||
            wcsncmp(new_path, L"//", 2) == 0)
            /* UNC path, nothing to do. */
            goto exit;
        
        env[1] = new_path[0];
        result = SetEnvironmentVariableW(env, new_path);

    exit:
        if (new_path != path_buf)
            PyMem_RawFree(new_path);
        return result;
    }

The single exit point half ensures the buffer is always freed. I also switched to using `MAXPATHLEN` instead of WinAPI `MAX_PATH`, in case we decide to increase its size later on.

izbyshev replied:
Thank you for your suggestion! I've considered using a loop here as in `win32_wgetcwd()`, but deemed `win32_wchdir()` too racy anyway: it may set a stale per-drive current directory in the environment, so I focused on fixing just the crash. But I reconsidered it given your comment and the opportunity to unify the code getting cwd. I've used your loop -- I like it.

eryksun replied:
I'm not terribly worried about undefined behavior when two threads compete over setting the working directory. Such code is already buggy. We could emit a warning, but it shouldn't fail.

eryksun commented on Modules/posixmodule.c:
```diff
@@ -3252,36 +3300,8 @@ posix_getcwd(int use_bytes)
-            PyMem_RawFree(wbuf2);
-        return resobj;
-    }
+    if (!use_bytes)
```
`os.getcwdb` is also incorrect. It shouldn't call the CRT's ANSI function. It needs to call `win32_wgetcwd` and convert the result to the file-system encoding, which defaults to UTF-8. This was apparently overlooked when implementing PEP 529.

izbyshev replied:
Indeed. According to PEP 529, `os.getcwdb` should have also been undeprecated. I'll file an issue for that in a while.

eryksun commented on Modules/posixmodule.c:
```diff
@@ -1508,36 +1541,28 @@ posix_fildes_fd(int fd, int (*func)(int))
 {
-    wchar_t path_buf[MAX_PATH], *new_path = path_buf;
-    int result;
+    wchar_t path_buf[MAXPATHLEN], *new_path;
```
`MAXPATHLEN` was changed to 256 in Windows. That's a strange value considering `MAX_PATH` is 260 and buffers sometimes need to be `MAX_PATH + 1` even with legacy code. 

Anyway, I don't think performance is critical for `getcwd()` and `chdir()`. It would be simpler to merge `win32_wgetcwd_buf()` into `win32_wgetcwd()`, and change the definition to `static wchar_t *win32_wgetcwd(LPDWORD plength)`. The caller would have to free the buffer.

izbyshev replied:
> `MAXPATHLEN` was changed to 256 in Windows. That's a strange value considering `MAX_PATH` is 260 and buffers sometimes need to be `MAX_PATH + 1` even with legacy code.

`MAXPATHLEN` is indeed strange, I even opened an issue about it:  #88822.

I believe I changed it purely for consistency with other code in posixmodule.c, but this might be the case of foolish consistency. I've now switched to `MAX_PATH` instead.

eryksun commented on Modules/posixmodule.c:
```diff
@@ -3242,6 +3267,28 @@ os_lchown_impl(PyObject *module, path_t *path, uid_t uid, gid_t gid)
+    if (wbuf2 != wbuf)
+        PyMem_RawFree(wbuf2);
+    return resobj;
+}
```
Here's the suggested change to make `win32_wgetcwd()` a function that doesn't hold the GIL and always allocates the buffer.

```suggestion
static wchar_t *
win32_wgetcwd(wchar_t *buf, size_t size)
{
    wchar_t *local_buf = buf;
    DWORD local_size = (DWORD)size;

    if (size > DWORD_MAX) {
        SetLastError(ERROR_INVALID_PARAMETER);
        errno = EINVAL;
        return NULL;
    }

    // If buf is NULL, allocate a buffer with the given size. If size is 0,
    // the result of GetCurrentDirectoryW() wil be used.
    if (buf == NULL && size > 0) {
        local_buf = PyMem_RawMalloc(dwsize * sizeof(wchar_t));
        if (local_buf == NULL) {
            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
            errno = ENOMEM;
            return NULL;
        }
    }

    while (1) {
        wchar_t *temp;
        DWORD result = GetCurrentDirectoryW(local_size, local_buf);
        if (!result) {
            // Set errno for the sake of completeness, but the result is never
            // 0 in practice. The call could fail due to an OS exception such
            // as an access violation, but that will just crash the process if
            // there's no exception handler. 
            errno = winerror_to_errno(GetLastError()); 
            goto fail;
        }

        // Note that the null terminator is not counted on success.
        if (result < local_size) {
            break;
        }

        // Reallocate only if buf is NULL and size is 0.
        if (buf != NULL || size != 0) {
            SetLastError(ERROR_INSUFFICIENT_BUFFER);
            errno = ERANGE;
            goto fail;
        }

        temp = PyMem_RawRealloc(local_buf, result * sizeof(wchar_t));
        if (!temp) {
            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
            errno = ENOMEM;
            goto fail;
        }
        local_buf = temp;
        local_size = result;
    }

    return local_buf;

fail:
    if (buf == NULL && local_buff != NULL) {
        PyMem_RawFree(local_buf);
    }
    return NULL;
}
```

eryksun replied:
I've updated the suggestion to implement `static wchar_t *win32_wgetcwd(wchar_t *buf, size_t size)` more faithfully to POSIX `getcwd()`. Like glibc, passing `buf` as `NULL` supports allocating the buffer. The caller is responsible for freeing an allocated buffer.

izbyshev replied:
I agree with the idea of inlining `win32_wgetcwd()` back into `posix_getcwd()`. In 4 years since my original PR handling of `use_bytes` was changed, and now it seems to make sense to have conversion from the native path type to Python's unicode/bytes in the same place for both Windows and Unix.

However, I can't see the value of introducing a pretty generic POSIX-like `win32_wgetcwd()` as you suggest. We need to get the current directory in just two places, and I don't see why our needs would change in the future.

I've force-pushed a new commit, which is mostly just a rebase of the original PR + inlining of `win32_wgetcwd()` and renaming of `win32_wgetcwd_buf()` to `win32_wgetcwd()`. ISTM that  `win32_wgetcwd()` is pretty simple. Having the caller to compare the returned buffer to the passed one is surely a wrinkle, but I don't think it matters, and overall I don't see why using your generic `win32_wgetcwd()` would be much better. But if there is some rationale that I'm missing, I'll be happy to revisit this.

eryksun replied:
> But if there is some rationale that I'm missing, I'll be happy to revisit this.

There's no deep rationale. I just figured I may as well copy the Linux behavior. POSIX [`getcwd()`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getcwd.html) only says that the behavior is "unspecified" when `buf` is `NULL`. Your code implements part of the extended behavior, given `buf` is `NULL` and `buf_size` is 0. However, `GetCurrentDirectoryW()` assumes that `nBufferLength` accurately represents the size of `lpBuffer`. If this size is big enough to hold the path of the current working directory, but `lpBuffer` is `NULL`, `GetCurrentDirectoryW()` will try to write to a `NULL` pointer and crash on an access violation. This can't happen in how we're currently using the function, but I felt it wouldn't hurt to round out the implementation. The Linux behavior is to allocate a new buffer in this case. Another option is to simply fail if `buf_size` is greater than 0 when `buf` is `NULL`.

izbyshev replied:
I see. I've intended `win32_getcwd()` purely as a helper for the two cases we have, not as a generic function (and one of those cases [shouldn't probably exist at all](https://github.com/python/cpython/pull/5802#issuecomment-1221621415)). I generally don't like to abstract over a small number of use cases. It works for `buf` is `NULL` and `buf_size` is 0 purely by accident. I can add an `assert()` or a failure case for `buf == NULL` if you like. 

I see how it could be tempting to add a more flexible version, but I think that would be justified if we needed to implement `Py_getcwd()`  for Python/fileutils.c, but not here.

eryksun commented on Modules/posixmodule.c:
```diff
@@ -3252,36 +3299,8 @@ posix_getcwd(int use_bytes)
-        return resobj;
-    }
+    if (!use_bytes)
+        return win32_wgetcwd();
```
With `win32_wgetcwd()` implemented as a lower-level function, more of the work remains in `posix_getcwd()`, but it's still relatively simple:

```suggestion
    PyObject *obj
    wchar_t *buf;

    Py_BEGIN_ALLOW_THREADS
    buf = win32_wgetcwd(NULL, 0);
    Py_END_ALLOW_THREADS

    if (!buf) {
        PyErr_NoMemory();
        return NULL;
    }

    obj = PyUnicode_FromWideChar(buf, -1);
    PyMem_RawFree(buf);

    if (use_bytes && obj) {
        Py_SETREF(obj, PyUnicode_EncodeFSDefault(obj));
    }

    return obj;
```

Commit:
gh-77085: Fix potential crash in os.chdir() and os.getcwd() on Windows

Both functions retry GetCurrentDirectoryW() if the initial buffer is
insufficient, but don't check whether the newly allocated buffer is
sufficient after the second call. This might not be the case if another
thread changes the current directory concurrently, and if so, the
functions will proceed to use uninitialized memory.

Fix this by retrying GetCurrentDirectoryW() in a loop with larger
buffers until it succeeds or we run out of memory.

Co-authored-by: Eryk Sun <eryksun@gmail.com>

eryksun commented on Modules/posixmodule.c:
```diff
@@ -1734,6 +1734,39 @@ posix_fildes_fd(int fd, int (*func)(int))
+        wchar_t *temp;
+        DWORD result = GetCurrentDirectoryW(buf_size, local_buf);
+        if (!result)
+            goto fail;
```
Per PEP 7, all new code requires braces, even where C allows omitting them.

```suggestion
        if (!result) {
            goto fail;
        }
```

izbyshev replied:
Ah, sorry about that, for some reason the braces were missing in the original PR too.

eryksun commented on Modules/posixmodule.c:
```diff
@@ -1734,6 +1734,39 @@ posix_fildes_fd(int fd, int (*func)(int))
+
+        /* L'\0' is not counted in result on success. */
+        if (result < buf_size)
+            break;
```
```suggestion
        if (result < buf_size) {
            break;
        }
```

izbyshev replied:
Fixed.

Commit:
Add braces per PEP 7

eryksun commented on Modules/posixmodule.c:
```diff
@@ -1734,6 +1734,39 @@ posix_fildes_fd(int fd, int (*func)(int))
+        temp = PyMem_RawRealloc(local_buf != buf ? local_buf : NULL,
+                                buf_size * sizeof(wchar_t));
+        if (!temp) {
+            SetLastError(ERROR_OUTOFMEMORY);
```
```suggestion
            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
```

The error code `ERROR_OUTOFMEMORY` is rarely seen in the Windows API, and only occurs when hard coded, likely by a Microsoft developer who picked the first error code that looked right instead of researching the error that was in common use. The automatic mapping from NT is `RtlNtStatusToDosError(STATUS_NO_MEMORY)` -> `ERROR_NOT_ENOUGH_MEMORY`. That's also the error code that the C runtime maps to the errno value `ENOMEM`, which Python defines as well in [`winerror_to_errno()`](https://github.com/python/cpython/blob/d8c7a1174cc182668085b10aab4049f6a2794c2f/PC/errmap.h#L71-L75) in "PC/errmap.h".

izbyshev replied:
Thanks for catching that. I thought I'd researched it when doing the original PR.

Commit:
ERROR_OUTOFMEMORY -> ERROR_NOT_ENOUGH_MEMORY
